File: GameLibrary/src/cc/lib/swing/AWTProbotLevelBuilder.java
package cc.lib.swing;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

import javax.swing.AbstractAction;
import javax.swing.ActionMap;
import javax.swing.InputMap;
import javax.swing.JFileChooser;
import javax.swing.JLabel;
import javax.swing.JScrollPane;
import javax.swing.JTextPane;
import javax.swing.KeyStroke;
import javax.swing.text.Keymap;

import cc.lib.game.GColor;
import cc.lib.game.Utils;
import cc.lib.probot.*;
import cc.lib.utils.FileUtils;
import cc.lib.utils.Grid;
import cc.lib.utils.Reflector;

public class AWTProbotLevelBuilder extends AWTComponent {

    public static void main(String [] args) throws Exception {
        Utils.DEBUG_ENABLED=true;
        new AWTProbotLevelBuilder();
    }

    final AWTFrame frame;
    final AWTRadioButtonGroup<Type> cellType;
    final AWTButton prev1, next1, prev10, next10;
    final JLabel levelNumLabel = new JLabel();
    final AWTToggleButton lazer0;
    final AWTToggleButton lazer1;
    final AWTToggleButton lazer2;
    final AWTNumberPicker npTurns, npJumps, npLoops;
    final AWTEditText levelLabel;
    final JTextPane info = new JTextPane();

    final Probot probot = new Probot() {
        @Override
        protected float getStrokeWidth() {
            return 5;
        }
    };
    final ArrayList<Level> levels = new ArrayList<>();
    int curLevel = 0;
    final File BACKUP_FILE;

    final String LEVEL_FILE = "levelsFile";

    AWTProbotLevelBuilder() throws Exception {
        final File settings = FileUtils.getOrCreateSettingsDirectory(getClass());
        BACKUP_FILE = new File(settings, "levels_backup.txt");
        frame = new AWTFrame() {

            @Override
            protected void onWindowClosing() {
                try {
                    Reflector.serializeToFile(levels, BACKUP_FILE);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        };

        {
            if (BACKUP_FILE.isFile()) {
                try {
                    levels.addAll(Reflector.deserializeFromFile(BACKUP_FILE));
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }

            if (levels.size() == 0) {
                levels.add(new Level());
            }
        }

        setMouseEnabled(true);

        frame.add(this);
        frame.addMenuBarMenu("File", new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent ev) {
                switch (ev.getActionCommand()) {
                    case "New": {
                        frame.setProperty(LEVEL_FILE, null);
                        levels.clear();
                        levels.add(new Level());
                        updateAll();
                        break;
                    }
                    case "Open": {
                        File toOpen = frame.showFileOpenChooser("Open Levels File", null);
                        if (toOpen != null) {
                            frame.log.info("Opening '%s'", toOpen);
                            try {
                                List<Level> newLevels = Reflector.deserializeFromFile(toOpen);
                                frame.setProperty(LEVEL_FILE, toOpen.getAbsolutePath());
                                levels.clear();
                                levels.addAll(newLevels);
                                updateAll();
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                        break;
                    }
                    case "Save":
                        try {
                            String levelsStr  = frame.getProperties().getProperty(LEVEL_FILE);
                            frame.log.info("Saving to '%s'",levelsStr);
                            if (levelsStr  == null) {
                                File levelsFile = frame.showFileSaveChooser("Choose File to save", null, null);
                                if (levelsFile != null) {
                                    Reflector.serializeToFile(levels, levelsFile);
                                    frame.setProperty(LEVEL_FILE, levelsFile.getAbsolutePath());
                                }
                            } else {
                                Reflector.serializeToFile(levels, new File(levelsStr));
                            }
                            updateAll();
                        } catch (Exception e) {
                            frame.setProperty(LEVEL_FILE, null);
                            e.printStackTrace();
                        }
                        break;
                    case "Save as": {
                        File levelsFile = null;
                        String str = frame.getProperties().getProperty(LEVEL_FILE);
                        if (str != null) {
                            levelsFile = new File(str);
                        }
                        File newFile = frame.showFileSaveChooser("Choose File to save", null, levelsFile);
                        if (newFile != null) {
                           frame.log.info("Saving to '%d'", newFile);
                            try {
                                Reflector.serializeToFile(levels, newFile);
                                frame.setProperty(LEVEL_FILE, newFile.getAbsolutePath());
                                updateAll();
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                        break;
                    }
                }
            }
        }, "New", "Open", "Save", "Save as");

        AWTPanel rhs = new AWTPanel(0, 1);
        cellType = new AWTRadioButtonGroup<Type>(rhs) {
            @Override
            protected void onChange(Type extra) {

            }
        };
        for (Type t : Type.values()) {
            cellType.addButton(t.displayName, t);
        }

        frame.add(rhs, BorderLayout.EAST);


        AWTPanel lhs = new AWTPanel();
        lhs.setLayout(new AWTButtonLayout(lhs));
        frame.add(lhs, BorderLayout.WEST);
        info.addKeyListener(new KeyListener() {
            @Override
            public void keyTyped(KeyEvent e) {
                levels.get(curLevel).info = info.getText();
            }

            @Override
            public void keyPressed(KeyEvent e) {
                levels.get(curLevel).info = info.getText();
            }

            @Override
            public void keyReleased(KeyEvent e) {
                levels.get(curLevel).info = info.getText();
            }
        });
        lhs.add(new AWTButton("Clear") {
            @Override
            protected void onAction() {
                grid.init(1, 1, Type.EM);
                getLevel().coins = grid.getGrid();
                AWTProbotLevelBuilder.this.repaint();
            }
        });

        lhs.add(new AWTButton("Reload"){
            @Override
            protected void onAction() {
                if (BACKUP_FILE.isFile()) {
                    try {
                        levels.clear();
                        levels.addAll(Reflector.deserializeFromFile(BACKUP_FILE));
                        updateAll();
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }
        });
        lhs.add(new AWTButton("Insert\nBefore") {
            @Override
            protected void onAction() {
                levels.add(curLevel, new Level());
                probot.setLevel(curLevel, levels.get(curLevel));
                updateAll();
                AWTProbotLevelBuilder.this.repaint();
            }
        });

        lhs.add(npTurns = new AWTNumberPicker.Builder().setMin(-1).setMax(10).setValue(-1).setLabel("Turns").build((int oldValue, int newValue) -> getLevel().numTurns = newValue ));
        lhs.add(npLoops = new AWTNumberPicker.Builder().setMin(-1).setMax(10).setValue(-1).setLabel("Loops").build((int oldValue, int newValue) -> getLevel().numLoops = newValue ));
        lhs.add(npJumps = new AWTNumberPicker.Builder().setMin(-1).setMax(10).setValue(-1).setLabel("Jumps").build((int oldValue, int newValue) -> getLevel().numJumps = newValue ));

        lhs.add(lazer0 = new AWTToggleButton("Lazer 0") {
            @Override
            protected void onToggle(boolean on) {
                getLevel().lazers[0] = on;
                probot.setLazerEnabled(0, on);
                AWTProbotLevelBuilder.this.repaint();
            }
        });
        lhs.add(lazer1 = new AWTToggleButton("Lazer 1") {
            @Override
            protected void onToggle(boolean on) {
                getLevel().lazers[1] = on;
                probot.setLazerEnabled(1, on);
                AWTProbotLevelBuilder.this.repaint();
            }
        });
        lhs.add(lazer2 = new AWTToggleButton("Lazer 2") {
            @Override
            protected void onToggle(boolean on) {
                getLevel().lazers[2] = on;
                probot.setLazerEnabled(2, on);
                AWTProbotLevelBuilder.this.repaint();
            }
        });
        info.setPreferredSize(new Dimension(100, 300));
        lhs.add(new JScrollPane(info));

        prev1 = new AWTButton("<") {
            @Override
            protected void onAction() {
                if (curLevel > 0) {
                    probot.setLevel(--curLevel, levels.get(curLevel));
                    updateAll();
                    AWTProbotLevelBuilder.this.repaint();
                }
            }
        };

        next1 = new AWTButton(">") {
            @Override
            protected void onAction() {
                if (curLevel == levels.size()-1) {
                    levels.add(new Level());
                }
                probot.setLevel(++curLevel, levels.get(curLevel));
                updateAll();
                AWTProbotLevelBuilder.this.repaint();
            }
        };

        prev10 = new AWTButton("<<") {
            @Override
            protected void onAction() {
                if (curLevel > 0) {
                    curLevel = Math.max(0, curLevel-10);
                    probot.setLevel(curLevel, levels.get(curLevel));
                    updateAll();
                    AWTProbotLevelBuilder.this.repaint();
                }
            }
        };

        next10 = new AWTButton(">>") {
            @Override
            protected void onAction() {
                if (curLevel < levels.size()-1) {
                    curLevel = Math.min(curLevel+10, levels.size()-1);
                    probot.setLevel(curLevel, levels.get(curLevel));
                    updateAll();
                    AWTProbotLevelBuilder.this.repaint();
                }
            }
        };

        levelLabel = new AWTEditText("", 32) {
            @Override
            protected void onTextChanged(String newText) {
                getLevel().label = newText;
            }
        };

        AWTPanel top = new AWTPanel(levelLabel, prev10, prev1, levelNumLabel, next1, next10);
        frame.add(top, BorderLayout.NORTH);

        if (!frame.loadFromFile(new File(settings, "probotlevelbuilder.properties")))
            frame.centerToScreen(800, 640);

        grid.setGrid(getLevel().coins);
        updateAll();

    }

    private Level getLevel() {
        return levels.get(curLevel);
    }

    private void updateAll() {
        levelNumLabel.setText("Level " + (curLevel+1) + " of " + levels.size());
        Level level = getLevel();
        grid.setGrid(level.coins);
        lazer0.setSelected(level.lazers[0]);
        lazer1.setSelected(level.lazers[1]);
        lazer2.setSelected(level.lazers[2]);
        prev1.setEnabled(curLevel > 0);
        prev10.setEnabled(curLevel > 0);
        next10.setEnabled(curLevel < levels.size()-1);
        npJumps.setValue(level.numJumps);
        npLoops.setValue(level.numLoops);
        npTurns.setValue(level.numTurns);
        levelLabel.setText(level.label);
        info.setText(level.info);
        String fileName = frame.getStringProperty(LEVEL_FILE, null);
        frame.setTitle(fileName == null ? "<UNNAMED>" : fileName);
    }

    int pickCol = -1;
    int pickRow = -1;

    Grid<Type> grid = new Grid() {
        @Override
        protected Type[][] build(int rows, int cols) {
            return new Type[rows][cols];
        }
    };

    @Override
    protected void paint(AWTGraphics g, int mouseX, int mouseY) {

        if (levels.size() == 0)
            return;
        // draw a grid

        final int cellDim = Math.min(g.getViewportWidth(), g.getViewportHeight()) / 20;
        final int viewWidth = g.getViewportWidth();
        final int viewHeight = g.getViewportHeight();

        probot.setLevel(curLevel, getLevel());
        probot.draw(g, grid.getCols()*cellDim, grid.getRows()*cellDim);
        g.setColor(GColor.WHITE);

        for (int i=0; i<=cellDim; i++) {
            g.drawLine(0, i*cellDim, viewWidth, i*cellDim);
            g.drawLine(i*cellDim, 0, i*cellDim, viewHeight);
        }

        pickCol = mouseX / cellDim;
        pickRow = mouseY / cellDim;

        //Utils.println("paint mx=" + mouseX + " my=" + mouseY);

        g.setColor(GColor.RED);

        g.drawRect(pickCol * cellDim, pickRow *cellDim, cellDim, cellDim);


    }

    @Override
    protected void onKeyPressed(VKKey key) {
        if (!grid.isValid(pickCol, pickRow))
            return;

        Type t = grid.get(pickRow, pickRow);
        Type [] values = {};

        switch (t) {

            case EM:
            case DD:
                values = Utils.toArray(Type.EM, Type.DD);
                break;
            case SE:
            case SS:
            case SW:
            case SN:
                values = Utils.toArray(Type.SE, Type.SS, Type.SW, Type.SN);
                break;
            case LH0:
            case LV0:
            case LH1:
            case LV1:
            case LH2:
            case LV2:
                values = Utils.toArray(Type.LH0, Type.LV0, Type.LH1, Type.LV1, Type.LH2, Type.LV2);
                break;
            case LB0:
            case LB1:
            case LB2:
            case LB:
                values = Utils.toArray(Type.LB0, Type.LB1, Type.LB2, Type.LB);
                break;
        }

        switch (key) {
            case VK_LEFT:
                t = Utils.decrementValue(t, values);
                break;
            case VK_RIGHT:
                t = Utils.incrementValue(t, values);
                break;
        }

        grid.set(pickRow, pickCol, t);
        repaint();
    }

    @Override
    protected void onClick() {
        if (pickCol < 0 || pickRow < 0) {
            return;
        }

        grid.ensureCapacity(pickRow+1, pickCol+1, Type.EM);
        getLevel().coins = grid.getGrid();

        Type t = this.cellType.getChecked();
        grid.set(pickRow, pickCol, t);

        repaint();
    }

    void drawGuy(AWTGraphics g, int x, int y, int radius, Direction dir) {
        g.drawFilledCircle(x, y, radius);
        g.setColor(GColor.BLACK);
        switch (dir) {
            case Right:
                g.drawLine(x, y, x+radius, y, 4);
                break;
            case Down:
                g.drawLine(x, y, x, y+radius, 4);
                break;
            case Left:
                g.drawLine(x, y, x-radius, y, 4);
                break;
            case Up:
                g.drawLine(x, y, x, y-radius, 4);
                break;
        }
    }
}

File: SwingApps/apps/cc/applets/probotbb/AWTProbotLevelBuilder.kt
package cc.applets.probotbb

import cc.lib.game.GColor
import cc.lib.game.Utils
import cc.lib.probot.Direction
import cc.lib.probot.Level
import cc.lib.probot.Probot
import cc.lib.probot.Type
import cc.lib.reflector.Reflector
import cc.lib.swing.*
import cc.lib.utils.FileUtils
import cc.lib.utils.Grid
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import java.io.File
import javax.swing.JLabel
import javax.swing.JScrollPane
import javax.swing.JTextPane

class AWTProbotLevelBuilder internal constructor() : AWTComponent() {
	val frame: AWTFrame
	val cellType: AWTRadioButtonGroup<Type>
	val prev1: AWTButton
	val next1: AWTButton
	val prev10: AWTButton
	val next10: AWTButton
	val levelNumLabel = JLabel()
	var lazer0: AWTToggleButton? = null
	var lazer1: AWTToggleButton? = null
	var lazer2: AWTToggleButton? = null
	var npTurns: AWTNumberPicker? = null
	var npJumps: AWTNumberPicker? = null
	var npLoops: AWTNumberPicker? = null
	val levelLabel: AWTEditText
	val info = JTextPane()
	val probot: Probot = object : Probot() {
		override fun getStrokeWidth(): Float {
			return 5f
		}
	}
	val levels = ArrayList<Level>()
	var curLevel = 0
	val BACKUP_FILE: File
	val LEVEL_FILE = "levelsFile"
	private val level: Level
		get() = levels[curLevel]

	private fun updateAll() {
		levelNumLabel.text = "Level " + (curLevel + 1) + " of " + levels.size
		val level = level
		grid.setGrid(level.coins)
		lazer0!!.isSelected = level.lazers[0]
		lazer1!!.isSelected = level.lazers[1]
		lazer2!!.isSelected = level.lazers[2]
		prev1.isEnabled = curLevel > 0
		prev10.isEnabled = curLevel > 0
		next10.isEnabled = curLevel < levels.size - 1
		npJumps!!.value = level.numJumps
		npLoops!!.value = level.numLoops
		npTurns!!.value = level.numTurns
		levelLabel.text = level.label
		info.text = level.info
		frame.title = frame.getStringProperty(LEVEL_FILE, "<UNNAMED>")
	}

	var pickCol = -1
	var pickRow = -1
	var grid: Grid<Type> = Grid<Type>()

	init {
		val settings = FileUtils.getOrCreateSettingsDirectory(javaClass)
		BACKUP_FILE = File(settings, "levels_backup.txt")
		frame = object : AWTFrame() {
			override fun onWindowClosing() {
				try {
					Reflector.serializeToFile<Any>(levels, BACKUP_FILE)
				} catch (e: Exception) {
					e.printStackTrace()
				}
			}
		}
		run {
			if (BACKUP_FILE.isFile) {
				try {
					levels.addAll(Reflector.deserializeFromFile(BACKUP_FILE))
				} catch (e: Exception) {
					e.printStackTrace()
				}
			}
			if (levels.size == 0) {
				levels.add(Level())
			}
			probot.setLevel(0, levels[0])
		}
		setMouseEnabled(true)
		frame.add(this)
		frame.addMenuBarMenu("File", { ev ->
			when (ev.actionCommand) {
				"New" -> {
					frame.setProperty(LEVEL_FILE, null)
					levels.clear()
					levels.add(Level())
					updateAll()
				}
				"Open" -> {
					val toOpen = frame.showFileOpenChooser("Open Levels File", "txt", null)
					if (toOpen != null) {
						frame.log.info("Opening '%s'", toOpen)
						try {
							val newLevels = Reflector.deserializeFromFile<List<Level>>(toOpen)
							frame.setProperty(LEVEL_FILE, toOpen.absolutePath)
							levels.clear()
							levels.addAll(newLevels)
							updateAll()
						} catch (e: Exception) {
							e.printStackTrace()
						}
					}
				}
				"Save" -> try {
					val levelsStr = frame.getProperties().getProperty(LEVEL_FILE)
					frame.log.info("Saving to '%s'", levelsStr)
					if (levelsStr == null) {
						val levelsFile = frame.showFileSaveChooser("Choose File to save", "Probot Levels", null, null)
						if (levelsFile != null) {
							Reflector.serializeToFile<Any>(levels, levelsFile)
							frame.setProperty(LEVEL_FILE, levelsFile.absolutePath)
						}
					} else {
						Reflector.serializeToFile<Any>(levels, File(levelsStr))
					}
					updateAll()
				} catch (e: Exception) {
					frame.setProperty(LEVEL_FILE, null)
					e.printStackTrace()
				}
				"Save as" -> {
					var levelsFile: File? = null
					val str = frame.getProperties().getProperty(LEVEL_FILE)
					if (str != null) {
						levelsFile = File(str)
					}
					val newFile = frame.showFileSaveChooser("Choose File to save", "Probot Levels", null, levelsFile)
					if (newFile != null) {
						frame.log.info("Saving to '%d'", newFile)
						try {
							Reflector.serializeToFile<Any>(levels, newFile)
							frame.setProperty(LEVEL_FILE, newFile.absolutePath)
							updateAll()
						} catch (e: Exception) {
							e.printStackTrace()
						}
					}
				}
			}
		}, "New", "Open", "Save", "Save as")
		val rhs = AWTPanel(0, 1)
		cellType = object : AWTRadioButtonGroup<Type>(rhs) {
			override fun onChange(extra: Type) {}
		}
		for (t in Type.values()) {
			cellType.addButton(t.displayName, t)
		}
		frame.add(rhs, BorderLayout.EAST)
		val lhs = AWTPanel()
		lhs.layout = AWTButtonLayout(lhs)
		frame.add(lhs, BorderLayout.WEST)
		info.addKeyListener(object : KeyListener {
			override fun keyTyped(e: KeyEvent) {
				levels[curLevel].info = info.text
			}

			override fun keyPressed(e: KeyEvent) {
				levels[curLevel].info = info.text
			}

			override fun keyReleased(e: KeyEvent) {
				levels[curLevel].info = info.text
			}
		})
		lhs.add(object : AWTButton("Clear") {
			override fun onAction() {
				grid.init(1, 1, Type.EM)
				level.coins = Array(grid.rows) { row ->
					Array(grid.cols) { col ->
						level.coins[row][col]
					}
				}
				this@AWTProbotLevelBuilder.repaint()
			}
		})
		lhs.add(object : AWTButton("Reload") {
			override fun onAction() {
				if (BACKUP_FILE.isFile) {
					try {
						levels.clear()
						levels.addAll(Reflector.deserializeFromFile(BACKUP_FILE))
						updateAll()
					} catch (e: Exception) {
						e.printStackTrace()
					}
				}
			}
		})
		lhs.add(object : AWTButton("Insert\nBefore") {
			override fun onAction() {
				levels.add(curLevel, Level())
				probot.setLevel(curLevel, levels[curLevel])
				updateAll()
				this@AWTProbotLevelBuilder.repaint()
			}
		})
		lhs.add(AWTNumberPicker.Builder().setMin(-1).setMax(10).setValue(-1).setLabel("Turns").build { _, newValue: Int -> level.numTurns = newValue }.also { npTurns = it })
		lhs.add(AWTNumberPicker.Builder().setMin(-1).setMax(10).setValue(-1).setLabel("Loops").build { _, newValue: Int -> level.numLoops = newValue }.also { npLoops = it })
		lhs.add(AWTNumberPicker.Builder().setMin(-1).setMax(10).setValue(-1).setLabel("Jumps").build { _, newValue: Int -> level.numJumps = newValue }.also { npJumps = it })
		lhs.add(object : AWTToggleButton("Lazer 0") {
			override fun onToggle(on: Boolean) {
				level.lazers[0] = on
				probot.setLazerEnabled(0, on)
				this@AWTProbotLevelBuilder.repaint()
			}
		}.also { lazer0 = it })
		lhs.add(object : AWTToggleButton("Lazer 1") {
			override fun onToggle(on: Boolean) {
				level.lazers[1] = on
				probot.setLazerEnabled(1, on)
				this@AWTProbotLevelBuilder.repaint()
			}
		}.also { lazer1 = it })
		lhs.add(object : AWTToggleButton("Lazer 2") {
			override fun onToggle(on: Boolean) {
				level.lazers[2] = on
				probot.setLazerEnabled(2, on)
				this@AWTProbotLevelBuilder.repaint()
			}
		}.also { lazer2 = it })
		info.preferredSize = Dimension(100, 300)
		lhs.add(JScrollPane(info))
		prev1 = object : AWTButton("<") {
			override fun onAction() {
				if (curLevel > 0) {
					probot.setLevel(--curLevel, levels[curLevel])
					updateAll()
					this@AWTProbotLevelBuilder.repaint()
				}
			}
		}
		next1 = object : AWTButton(">") {
			override fun onAction() {
				if (curLevel == levels.size - 1) {
					levels.add(Level())
				}
				probot.setLevel(++curLevel, levels[curLevel])
				updateAll()
				this@AWTProbotLevelBuilder.repaint()
			}
		}
		prev10 = object : AWTButton("<<") {
			override fun onAction() {
				if (curLevel > 0) {
					curLevel = Math.max(0, curLevel - 10)
					probot.setLevel(curLevel, levels[curLevel])
					updateAll()
					this@AWTProbotLevelBuilder.repaint()
				}
			}
		}
		next10 = object : AWTButton(">>") {
			override fun onAction() {
				synchronized(this@AWTProbotLevelBuilder) {
					if (curLevel < levels.size - 1) {
						curLevel = Math.min(curLevel + 10, levels.size - 1)
						probot.setLevel(curLevel, levels[curLevel])
						updateAll()
						this@AWTProbotLevelBuilder.repaint()
					}
				}
			}
		}
		levelLabel = object : AWTEditText("", 32) {
			override fun onTextChanged(newText: String) {
				level.label = newText
			}
		}
		val top = AWTPanel(levelLabel, prev10, prev1, levelNumLabel, next1, next10)
		frame.add(top, BorderLayout.NORTH)
		if (!frame.loadFromFile(File(settings, "probotlevelbuilder.properties"))) frame.centerToScreen(800, 640)
		grid.setGrid(level.coins)
		updateAll()
	}

	@Synchronized
	override fun paint(g: AWTGraphics, mouseX: Int, mouseY: Int) {
		if (levels.size == 0) return
		// draw a grid
		val cellDim = Math.min(g.viewportWidth, g.viewportHeight) / 20
		val viewWidth = g.viewportWidth
		val viewHeight = g.viewportHeight

//        probot.setLevel(curLevel, getLevel());
		probot.draw(g, grid.cols * cellDim, grid.rows * cellDim)
		g.color = GColor.WHITE
		for (i in 0..cellDim) {
			g.drawLine(0f, (i * cellDim).toFloat(), viewWidth.toFloat(), (i * cellDim).toFloat())
			g.drawLine((i * cellDim).toFloat(), 0f, (i * cellDim).toFloat(), viewHeight.toFloat())
		}
		pickCol = mouseX / cellDim
		pickRow = mouseY / cellDim

		//Utils.println("paint mx=" + mouseX + " my=" + mouseY);
		g.color = GColor.RED
		g.drawRect((pickCol * cellDim).toFloat(), (pickRow * cellDim).toFloat(), cellDim.toFloat(), cellDim.toFloat())
	}

	@Synchronized
	override fun keyPressed(evt: KeyEvent) {
		if (!grid.isValid(pickCol, pickRow)) return
		var t = grid[pickRow, pickRow]
		var values = arrayOf<Type?>()
		when (t) {
			Type.EM, Type.DD -> values = Utils.toArray(Type.EM, Type.DD)
			Type.SE, Type.SS, Type.SW, Type.SN -> values = Utils.toArray(Type.SE, Type.SS, Type.SW, Type.SN)
			Type.LH0, Type.LV0, Type.LH1, Type.LV1, Type.LH2, Type.LV2 -> values = Utils.toArray(Type.LH0, Type.LV0, Type.LH1, Type.LV1, Type.LH2, Type.LV2)
			Type.LB0, Type.LB1, Type.LB2, Type.LB -> values = Utils.toArray(Type.LB0, Type.LB1, Type.LB2, Type.LB)
		}
		when (evt.keyCode) {
			KeyEvent.VK_LEFT -> t = Utils.decrementValue(t, *values)
			KeyEvent.VK_RIGHT -> t = Utils.incrementValue(t, *values)
		}
		grid[pickRow, pickCol] = t
		repaint()
	}

	override fun onClick() {
		if (pickCol < 0 || pickRow < 0) {
			return
		}
		grid.ensureCapacity(pickRow + 1, pickCol + 1, Type.EM)
		level.coins = Array(grid.rows) { row ->
			Array(grid.cols) { col ->
				grid.get(row, col)
			}
		}
		cellType.getChecked()?.let {
			grid[pickRow, pickCol] = cellType.getChecked()
		}
		repaint()
	}

	fun drawGuy(g: AWTGraphics, x: Int, y: Int, radius: Int, dir: Direction) {
		g.drawFilledCircle(x, y, radius)
		g.color = GColor.BLACK
		when (dir) {
			Direction.Right -> g.drawLine(x.toFloat(), y.toFloat(), (x + radius).toFloat(), y.toFloat(), 4f)
			Direction.Down -> g.drawLine(x.toFloat(), y.toFloat(), x.toFloat(), (y + radius).toFloat(), 4f)
			Direction.Left -> g.drawLine(x.toFloat(), y.toFloat(), (x - radius).toFloat(), y.toFloat(), 4f)
			Direction.Up -> g.drawLine(x.toFloat(), y.toFloat(), x.toFloat(), (y - radius).toFloat(), 4f)
		}
	}

	companion object {
		@Throws(Exception::class)
		@JvmStatic
		fun main(args: Array<String>) {
			Utils.setDebugEnabled()
			AWTProbotLevelBuilder()
		}
	}
}